mesonbuild/cargo/interpreter.py

```diff
@@ -0,0 +1,449 @@
+        raw_manifest.get('features', {}),
+        {k: {k2: Dependency.from_raw(v2) for k2, v2 in v['dependencies'].items()}
+         for k, v in raw_manifest.get('target', {}).items()},
+        subdir,
+        path,
+    )
+
+
+def _load_manifests(subdir: str) -> T.Dict[str, Manifest]:
+    filename = os.path.join(subdir, 'Cargo.toml')
+    raw = load_toml(filename)
+
+    manifests: T.Dict[str, Manifest] = {}
+
+    raw_manifest: T.Union[manifest.Manifest, manifest.VirtualManifest]
+    if 'package' in raw:
+        raw_manifest = T.cast('manifest.Manifest', raw)
+        manifest_ = _convert_manifest(raw_manifest, subdir)
+        manifests[manifest_.package.name] = manifest_
+    else:
+        raw_manifest = T.cast('manifest.VirtualManifest', raw)
+
+    if 'workspace' in raw_manifest:
+        # XXX: need to verify that python glob and cargo globbing are the
+        # same and probably write  a glob implementation. Blarg
+
+        # We need to chdir here to make the glob work correctly
+        pwd = os.getcwd()
+        os.chdir(subdir)
+        members: T.Iterable[str]
+        try:
+            members = itertools.chain.from_iterable(
+                glob.glob(m) for m in raw_manifest['workspace']['members'])
+        finally:
+            os.chdir(pwd)
+        if 'exclude' in raw_manifest['workspace']:
+            members = (x for x in members if x not in raw_manifest['workspace']['exclude'])
+
+        for m in members:
+            filename = os.path.join(subdir, m, 'Cargo.toml')
+            raw = load_toml(filename)
+
+            raw_manifest = T.cast('manifest.Manifest', raw)
+            man = _convert_manifest(raw_manifest, subdir, m)
+            manifests[man.package.name] = man
+
+    return manifests
+
+
+def load_all_manifests(subproject_dir: str) -> T.Dict[str, Manifest]:
+    """Find all cargo subprojects, and load them
+
+    :param subproject_dir: Directory to look for subprojects in
+    :return: A dictionary of rust project names to Manifests
+    """
+    manifests: T.Dict[str, Manifest] = {}
+    for p in Path(subproject_dir).iterdir():
+        if p.is_dir() and (p / 'Cargo.toml').exists():
+            manifests.update(_load_manifests(str(p)))
+    return manifests
+
+
+def _create_lib(cargo: Manifest, build: builder.Builder) -> T.List[mparser.BaseNode]:
+    kw: T.Dict[str, mparser.BaseNode] = {}
+    if cargo.dependencies:
+        ids = [build.identifier(f'dep_{n}') for n in cargo.dependencies]
+        kw['dependencies'] = build.array(
+            [build.method('get_variable', i, [build.string('dep')]) for i in ids])
+    kw['link_with'] = build.identifier('lib')
+
+    # FIXME: currently assuming that an rlib is being generated, which is
+    # the most common.
+    return [
+        build.assign(
+            build.function(
+                'static_library',
+                [
+                    build.string(fixup_meson_varname(cargo.package.name)),
```

Comment:
I don't think we need to replace `-` in library name here.

Reply:
Pretty sure rust targets can't have `-` in their outputs

Reply:
I thought Meson was already replacing dash to underscore when needed, but I was wrong: https://github.com/mesonbuild/meson/blob/master/mesonbuild/build.py#L2038. Note that dash is allowed for cdylib and staticlib, but we don't support those yet anyway. Let's ignore this then, you're right.